hyper/http20/connection.py

```diff
@@ -266,8 +266,9 @@ def request(self, method, url, body=None, headers=None):
         # If threads interleave these operations, it could result in messages
         # being sent in the wrong order, which can lead to the out-of-order
         # messages with lower stream IDs being closed prematurely.
+        stream_id = self.putrequest(method, url)
+        self.connect()
```

Comment:
I think the single lock model is likely to be the most usable approach at this time.